services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/crt/AbstractS3CrtAsyncClient.java

```diff
@@ -0,0 +1,806 @@
+import software.amazon.awssdk.services.s3.model.PutBucketRequestPaymentRequest;
+import software.amazon.awssdk.services.s3.model.PutBucketRequestPaymentResponse;
+import software.amazon.awssdk.services.s3.model.PutBucketTaggingRequest;
+import software.amazon.awssdk.services.s3.model.PutBucketTaggingResponse;
+import software.amazon.awssdk.services.s3.model.PutBucketVersioningRequest;
+import software.amazon.awssdk.services.s3.model.PutBucketVersioningResponse;
+import software.amazon.awssdk.services.s3.model.PutBucketWebsiteRequest;
+import software.amazon.awssdk.services.s3.model.PutBucketWebsiteResponse;
+import software.amazon.awssdk.services.s3.model.PutObjectAclRequest;
+import software.amazon.awssdk.services.s3.model.PutObjectAclResponse;
+import software.amazon.awssdk.services.s3.model.PutObjectLegalHoldRequest;
+import software.amazon.awssdk.services.s3.model.PutObjectLegalHoldResponse;
+import software.amazon.awssdk.services.s3.model.PutObjectLockConfigurationRequest;
+import software.amazon.awssdk.services.s3.model.PutObjectLockConfigurationResponse;
+import software.amazon.awssdk.services.s3.model.PutObjectRequest;
+import software.amazon.awssdk.services.s3.model.PutObjectResponse;
+import software.amazon.awssdk.services.s3.model.PutObjectRetentionRequest;
+import software.amazon.awssdk.services.s3.model.PutObjectRetentionResponse;
+import software.amazon.awssdk.services.s3.model.PutObjectTaggingRequest;
+import software.amazon.awssdk.services.s3.model.PutObjectTaggingResponse;
+import software.amazon.awssdk.services.s3.model.PutPublicAccessBlockRequest;
+import software.amazon.awssdk.services.s3.model.PutPublicAccessBlockResponse;
+import software.amazon.awssdk.services.s3.model.RestoreObjectRequest;
+import software.amazon.awssdk.services.s3.model.RestoreObjectResponse;
+import software.amazon.awssdk.services.s3.model.SelectObjectContentRequest;
+import software.amazon.awssdk.services.s3.model.SelectObjectContentResponseHandler;
+import software.amazon.awssdk.services.s3.model.UploadPartCopyRequest;
+import software.amazon.awssdk.services.s3.model.UploadPartCopyResponse;
+import software.amazon.awssdk.services.s3.model.UploadPartRequest;
+import software.amazon.awssdk.services.s3.model.UploadPartResponse;
+import software.amazon.awssdk.services.s3.model.WriteGetObjectResponseRequest;
+import software.amazon.awssdk.services.s3.model.WriteGetObjectResponseResponse;
+import software.amazon.awssdk.services.s3.paginators.ListMultipartUploadsPublisher;
+import software.amazon.awssdk.services.s3.paginators.ListObjectVersionsPublisher;
+import software.amazon.awssdk.services.s3.paginators.ListObjectsV2Publisher;
+import software.amazon.awssdk.services.s3.paginators.ListPartsPublisher;
+import software.amazon.awssdk.services.s3.waiters.S3AsyncWaiter;
+
+@SdkInternalApi
+public abstract class AbstractS3CrtAsyncClient implements S3CrtAsyncClient {
+    abstract S3AsyncClient s3AsyncClient();
```

Comment:
Per offline discussion, should be passed as a c-tor param

Reply:
Yeah, this is the first thing I tried when I was implementing it, but it doesn't work here because we need to instantiate `S3AsyncClient` first in the ctor of `DefaultS3CrtAsyncClient` (before `super(client)`). Another option is to use composition over inheritance, but we'd need to write the same amount of operations in the `DefaultS3CrtAsyncClient`

Reply:
To close the loop, we decided to use `super(someStaticMethodToInitializeS3Client())`